safekeeper/src/metrics.rs

```diff
@@ -63,13 +66,120 @@ pub static PERSIST_CONTROL_FILE_SECONDS: Lazy<Histogram> = Lazy::new(|| {
 });
 pub static PG_IO_BYTES: Lazy<IntCounterVec> = Lazy::new(|| {
     register_int_counter_vec!(
-        "safekeeper_pg_io_bytes",
+        "safekeeper_pg_io_bytes_total",
         "Bytes read from or written to any PostgreSQL connection",
-        &["direction"]
+        &["client_az", "sk_az", "app_name", "dir", "same_az"]
     )
     .expect("Failed to register safekeeper_pg_io_bytes gauge")
 });
 
+pub const LABEL_UNKNOWN: &str = "unknown";
+
+/// Labels for traffic metrics.
+struct ConnectionLabels {
+    /// Availability zone of the connection origin.
+    client_az: String,
+    /// Availability zone of the current safekeeper.
+    sk_az: String,
+    /// Client application name.
+    app_name: String,
+    /// Total bytes read from this connection.
+    read: GenericCounter<metrics::core::AtomicU64>,
+    /// Total bytes written to this connection.
+    write: GenericCounter<metrics::core::AtomicU64>,
+}
+
+impl ConnectionLabels {
+    fn new() -> Self {
+        let mut this = Self {
+            client_az: LABEL_UNKNOWN.to_string(),
+            sk_az: LABEL_UNKNOWN.to_string(),
+            app_name: LABEL_UNKNOWN.to_string(),
+            // Fill in the metrics with dummy values, they will be updated later.
+            read: GenericCounter::new("_", "_").unwrap(),
+            write: GenericCounter::new("_", "_").unwrap(),
+        };
+        this.update_metrics();
```

Comment:
This does look a bit weird...

If instead the counter creation was a function of these metric parameters, then it could easily be reused on update and in creation, not to create any of these invalid values. It doesn't seem that these would get registered when using `::new`.

Reply:
Created a `TrafficMetricsState` struct to avoid invalid values in constructor.